module2/lessons/data_validations.md

```diff
@@ -0,0 +1,93 @@
+
+Choose these in certain, specific cases. It's the opinion of the Rails team that model-level validations are the most appropriate in most circumstances.
+
+There are two kinds of Active Record objects: those that correspond to a row inside your database and those that do not. When you create a fresh object, for example using the new method, that object does not belong to the database yet. Once you call save upon that object it will be saved into the appropriate database table. Active Record uses the new_record? instance method to determine whether an object is already in the database or not.
+
+**app/models/artist.rb**
+
+```ruby
+class Artist < ApplicationRecord
+  has_many :songs
+
+  def average_song_length
+    self.songs.average(:length)
+  end
+end
+```
+
+If you were to look in the rails console you would see the following....
+
+```bash
+irb(main):001:0> a = Artist.new(name: "Carly Rae Jepsen")
+=> #<Artist:0x000000010abd7fd0 id: nil, name: "Carly Rae Jepsen", created_at: nil, updated_at: nil>
+irb(main):002:0> a.new_record?
+=> true
+# Not saved to the database yet, no validations have run. Notice the nil values.
+irb> a.save
+irb(main):003:0> a.save
+  TRANSACTION (0.4ms)  BEGIN
+  Artist Create (1.2ms)  INSERT INTO "artists" ("name", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"  [["name", "Carly Rae Jepsen"], ["created_at", "2023-02-23 16:48:17.820040"], ["updated_at", "2023-02-23 16:48:17.820040"]]
+  TRANSACTION (2.8ms)  COMMIT
+=> true
+# Validations were checked as part of the process for inserting a new record into the database.
+# Looking at our object now:
+irb(main):004:0> a
+=> #<Artist:0x000000010abd7fd0 id: 8, name: "Carly Rae Jepsen", created_at: Thu, 23 Feb 2023 16:48:17.820040000 UTC +00:00, updated_at: Thu, 23 Feb 2023 16:48:17.820040000 UTC +00:00>
+irb(main):005:0> a.new_record?
+=> false
+# This object now corresponds with a row in the database
+```
+
+Creating and saving a new record will send an SQL INSERT operation to the database. Updating an existing record will send an SQL UPDATE operation instead. **Validations are typically run before these commands are sent to the database.** If any validations fail, the object will be marked as invalid and Active Record will not perform the INSERT or UPDATE operation. This avoids storing an invalid object in the database. You can choose to have specific validations run when an object is created, saved, or updated with the `:on` option.
+
+The following methods trigger validations, and will save the object to the database only if the object is valid:
+
+- create
+- create!
+- save
+- save!
+- update
+- update!
+
+**The bang versions (e.g. save!) raise an exception if the record is invalid. The non-bang versions don't: save and update return false, and create returns the object.**
+
```

Comment:
Realizing this is a gap in the exising lesson, but I think it would be great to introduce ActiveRecord Validations here. We could add something like this:

So how do we tell Rails how to validate our models? We can use [ActiveRecord validations](https://guides.rubyonrails.org/active_record_validations.html) in our model classes to be explicit about what Rails should consider a valid or invalid record. To require an artist's name, we could do the following:
```ruby
class Artist < ApplicationRecord
  has_many :songs
  validates :name, presence: true
```

Try it < SECTION>
Once you add this validation, open up a Rails console and check out the differences when between AR methods when validations are present and an artist is created with no name:
`Artist.create()`
`Artist.create!()`
`Artist.new()`
`Artist.save`

Reply:
This is something we've had them do in class, just adding it here now since this will be async homework!

Reply:
Got it! Will explicitly introduce and link to info for ActiveRecord validations.